public/app/features/alerting/unified/rule-list/AlertRuleLoader.tsx

```diff
@@ -0,0 +1,122 @@
+import { memo, useMemo } from 'react';
+
+import { DataSourceRuleGroupIdentifier, Rule, RuleIdentifier } from 'app/types/unified-alerting';
+
+import { alertRuleApi } from '../api/alertRuleApi';
+import { featureDiscoveryApi } from '../api/featureDiscoveryApi';
+import { fromRule, equal, fromRulerRule, stringifyIdentifier } from '../utils/rule-id';
+import { getRulePluginOrigin, isAlertingRule, isRecordingRule } from '../utils/rules';
+import { createRelativeUrl } from '../utils/url';
+
+import { AlertRuleListItem, RecordingRuleListItem, UnknownRuleListItem } from './components/AlertRuleListItem';
+import { ActionsLoader, RuleActionsButtons } from './components/RuleActionsButtons.V2';
+
+const { useDiscoverDsFeaturesQuery } = featureDiscoveryApi;
+const { useGetRuleGroupForNamespaceQuery } = alertRuleApi;
+
+interface AlertRuleLoaderProps {
+  rule: Rule;
+  groupIdentifier: DataSourceRuleGroupIdentifier;
+}
+
+export const AlertRuleLoader = memo(function AlertRuleLoader({ rule, groupIdentifier }: AlertRuleLoaderProps) {
```

Comment:
why do we need this memo?

Reply:
Performance optimisation. We might render hundreds or thousands of these components and often when rerendering (adding more search results) we don't update existing components, we only want to render the new ones, so this memoization can greatly reduce CPU workload

Reply:
But then, I think we should use a custom comparison function in the memo. Anytime rules are set, a new array is created and the reference is updated, so the memo would not prevent the rerender in this case 🤔
```
 {rules.map(({ ruleKey, rule, groupIdentifier }) => (
          <AlertRuleLoader key={ruleKey} rule={rule} groupIdentifier={groupIdentifier} />
        ))}
```
in here rule would be a different reference any time we change rules, and memo without the custom function wouldn't prevent rerendering.

Reply:
New array is created, but most objects in the array are the same. I did check that using React Profiler some time ago, but I can double check